contracts/campaigns/KOLReferral.sol

```diff
@@ -0,0 +1,140 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import "@opengsn/contracts/src/ERC2771Recipient.sol";
+
+contract KOLReferral is Initializable, ERC2771Recipient {
+    address public admin;
+
+    mapping(address => string) public walletToReferralId;
+    mapping(string => address) public referralIdToWallet;
+    mapping(address => string) public userReferredBy;
+    mapping(address => address[]) private _kolToUsers;
+
+    address[] private _users;
+    address[] private _kols;
+
+    modifier onlyAdmin() {
+        require(_msgSender() == admin, "Only Admin");
+        _;
+    }
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() {
+        _disableInitializers();
+    }
+
+    function initialize(address admin_, address trustedForwarder_)
+        external
+        initializer
+    {
+        require(admin_ != address(0), "zero address");
+        require(trustedForwarder_ != address(0), "zero address");
+        admin = admin_;
+        _setTrustedForwarder(trustedForwarder_);
+    }
+
+    function registerKOL(address wallet, string memory referralId)
+        external
+        onlyAdmin
+    {
+        require(
+            referralIdToWallet[referralId] == address(0),
+            "ReferralId is already taken"
+        );
+        require(
+            bytes(walletToReferralId[wallet]).length == 0,
+            "ReferralId is already assigned for this wallet"
+        );
+        walletToReferralId[wallet] = referralId;
+        referralIdToWallet[referralId] = wallet;
+        _kols.push(wallet);
+    }
+
+    // TODO: use msgSender() for gasless transaction
```

Comment:
It is already msgSender right?

Reply:
ohh yes. Forgot to remove the comment.
Will remove the comment